### PR TITLE
feat(plugin-meetings): password and captcha support for unified space meetings

### PR DESCRIPTION
**Pull request by marcin-bazyl**
Initial draft of changes required to support password and captcha when talking with WbxAppApi

It's still got lots of todo comments, console logs, etc.

Tests need to be added/updated.
Some of the new code may be also be gated on the config setting for using wbxappapi.

### CODE REVIEW

**Comment by arun3528** [comment]
Nice start thanks @marcin-bazyl

**Comment by marcin-bazyl** [comment]
This is the implementation for part of the work described here: https://confluence-eng-gpk2.cisco.com/conf/display/WBXT/Webex+Password+Flow+For+SDK

File: packages/node_modules/@webex/plugin-meetings/src/meeting/index.js
Comment (arun3528) [inline packages/node_modules/@webex/plugin-meetings/src/meeting/index.js:942]: if you pass a valid captcha and bad password it still markes it as failure 
?

File: packages/node_modules/@webex/plugin-meetings/src/meeting/index.js
Comment (arun3528) [inline packages/node_modules/@webex/plugin-meetings/src/meeting/index.js:978]: looks good

File: packages/node_modules/@webex/plugin-meetings/src/meeting/index.js
Comment (arun3528) [inline packages/node_modules/@webex/plugin-meetings/src/meeting/index.js:1002]: Looks good

**Review by arun3528 — COMMENTED** [review]
Most of the things looks goof with minor changes and questions

File: packages/node_modules/@webex/plugin-meetings/src/constants.js
Comment (marcin-bazyl) [reply packages/node_modules/@webex/plugin-meetings/src/constants.js:1266]: yes, existing sdk code calls fetchMeetingInfo() immediately after the meeting object is created, but it is an async operation so someone might try to read the password status before fetchMeetingInfo() is completed. There are also other possibilities:
1. fetchMeetingInfo() may fail for some network/timeout reason
2. fetchMeetingInfo() may require captcha just because it was too many attempts to fetch info for non-existing meeting (code 423001)
in these 2 cases we will not know if password is required or not until we successfully try again

File: packages/node_modules/@webex/plugin-meetings/src/meeting-info/meeting-info-v2.js
Comment (marcin-bazyl) [reply packages/node_modules/@webex/plugin-meetings/src/meeting-info/meeting-info-v2.js:17]: this one is thrown and caught internally inside SDK, not thrown outside to the developer. It's specific to the meeting-info-v2 module so, if you want I can add a text "fetchMeetingInfo() has to be called again with the password password parameter"

File: packages/node_modules/@webex/plugin-meetings/src/meeting-info/meeting-info-v2.js
Comment (marcin-bazyl) [reply packages/node_modules/@webex/plugin-meetings/src/meeting-info/meeting-info-v2.js:104]: There are a few error codes and yeah initially I've put them all here, but then I've found the equivalent UCF source code and they don't check the error codes in UCF, only the http code like 403 or 423 so I thought it will be better to align with UCF on this (it means that if wbxappapi add 1 more 403xxxx code for something we won't have to change the SDK code).

File: packages/node_modules/@webex/plugin-meetings/src/meeting-info/meeting-info-v2.js
Comment (marcin-bazyl) [reply packages/node_modules/@webex/plugin-meetings/src/meeting-info/meeting-info-v2.js:107]: see my answer above

File: packages/node_modules/@webex/plugin-meetings/src/meeting/index.js
Comment (marcin-bazyl) [reply packages/node_modules/@webex/plugin-meetings/src/meeting/index.js:862]: see my earlier answer for this

File: packages/node_modules/@webex/plugin-meetings/src/meeting/index.js
Comment (marcin-bazyl) [reply packages/node_modules/@webex/plugin-meetings/src/meeting/index.js:876]: we're setting it to null once verified

File: packages/node_modules/@webex/plugin-meetings/src/meeting/index.js
Comment (marcin-bazyl) [reply packages/node_modules/@webex/plugin-meetings/src/meeting/index.js:886]: It is required by the web client (we need to show only 1 error - either for password or for captcha). Wbxappapi has some bugs/limitations and having this field makes it easier for the web app to know what caused the failure (captcha or password). Also in cases where something else fails, this field will give you an indication why meetingInfo is not available (although currently I've only put 1 value for that: "other", but it can be expanded). So I've added it thinking that if it's useful for web client then it will probably be useful for others. I can make it private for now and keep using it in the web client or I can remove it completely from the SDK and put that logic in the web client if you prefer.

File: packages/node_modules/@webex/plugin-meetings/src/meeting/index.js
Comment (marcin-bazyl) [reply packages/node_modules/@webex/plugin-meetings/src/meeting/index.js:942]: yes, it's a bug in wbxappapi: when you give them a good captcha with bad password they don't ask for captcha anymore, but they won't give you the meeting info even if you provide the good password in next attempt. The only way to make it work is to ask for another captcha - same workaround was done in UCF

File: packages/node_modules/@webex/plugin-meetings/src/constants.js
Comment (marcin-bazyl) [reply packages/node_modules/@webex/plugin-meetings/src/constants.js]: done

File: packages/node_modules/@webex/plugin-meetings/src/constants.js
Comment (marcin-bazyl) [reply packages/node_modules/@webex/plugin-meetings/src/constants.js]: done

File: packages/node_modules/@webex/plugin-meetings/src/meeting-info/meeting-info-v2.js
Comment (marcin-bazyl) [reply packages/node_modules/@webex/plugin-meetings/src/meeting-info/meeting-info-v2.js]: done

**Comment by marcin-bazyl** [comment]
I've done PR suggestions as a new commit so it's easier to see what's changed. Once PR is approved I'll squash all commits into 1 for merging with master

File: packages/node_modules/@webex/plugin-meetings/src/meeting-info/meeting-info-v2.js
Comment (arun3528) [reply packages/node_modules/@webex/plugin-meetings/src/meeting-info/meeting-info-v2.js:104]: If all the scenarios are working fine then yes , But i dont trust these guys

**Comment by arun3528** [comment]
Looks good to me

**Review by arun3528 — APPROVED** [review]

**Review by taymoork2 — APPROVED** [review]
LGTM

My only small gripe would be to make the default messages in both `MeetingInfoV2...Error` constructors constants

**Comment by marcin-bazyl** [comment]
the last force push is just a squash of all the commits into 1 commit so that it's ready for merging plus 2 small changes:
1. making `meeting.meetingInfoFailureReason` private as suggested by Arun in 1 of the comments
2. adding constants for `MeetingInfoV2*Error` default messages as suggested by Taymoor (this change can be seen on its own in commit 1f34043)

File: packages/node_modules/@webex/plugin-meetings/src/meeting/index.js
Comment (lalli-flores) [reply packages/node_modules/@webex/plugin-meetings/src/meeting/index.js:942]: @marcin-bazyl do you have a follow up ticket to revisit this when it gets fixed in the API?

File: packages/node_modules/@webex/plugin-meetings/src/meeting-info/meeting-info-v2.js
Comment (lalli-flores) [inline packages/node_modules/@webex/plugin-meetings/src/meeting-info/meeting-info-v2.js:6]:
```suggestion
const PASSWORD_ERROR_DEFAULT_MESSAGE = 'Password required. Call fetchMeetingInfo() with password argument';
```

File: packages/node_modules/@webex/plugin-meetings/src/meeting-info/meeting-info-v2.js
Comment (lalli-flores) [inline packages/node_modules/@webex/plugin-meetings/src/meeting-info/meeting-info-v2.js:7]:
```suggestion
const CAPTCHA_ERROR_DEFAULT_MESSAGE = 'Captcha required. Call fetchMeetingInfo() with captchaInfo argument';
```

File: packages/node_modules/@webex/plugin-meetings/src/meeting/index.js
Comment (lalli-flores) [inline packages/node_modules/@webex/plugin-meetings/src/meeting/index.js:905]: Can you make this error more specific? when looking at the logs this is not helpful

File: packages/node_modules/@webex/plugin-meetings/src/meeting/index.js
Comment (lalli-flores) [inline packages/node_modules/@webex/plugin-meetings/src/meeting/index.js:908]: Can you make this error more specific? when looking at the logs this is not helpful

File: packages/node_modules/@webex/plugin-meetings/src/meeting/index.js
Comment (marcin-bazyl) [reply packages/node_modules/@webex/plugin-meetings/src/meeting/index.js]: done

File: packages/node_modules/@webex/plugin-meetings/src/meeting/index.js
Comment (marcin-bazyl) [reply packages/node_modules/@webex/plugin-meetings/src/meeting/index.js]: done

File: packages/node_modules/@webex/plugin-meetings/src/meeting/index.js
Comment (marcin-bazyl) [reply packages/node_modules/@webex/plugin-meetings/src/meeting/index.js:942]: ok, I've raised https://jira-eng-gpk2.cisco.com/jira/browse/SPARK-293259 for this, but don't hold your breath for the wbxappapi fix.... ;-)

File: packages/node_modules/@webex/plugin-meetings/src/meeting/request.js
Comment (arun3528) [inline packages/node_modules/@webex/plugin-meetings/src/meeting/request.js:150]: Missed this , Can we move this to the meeting/request  in this or the next PR

**Comment by marcin-bazyl** [comment]
It looks like the server team has done a fix for WEBEX-224862, for which I have a workaround in this PR, so we should hold off from merging this PR until I confirm with them where that fix has been deployed and check that all use cases work correctly with their new implementation.

**Review by arun3528 — CHANGES_REQUESTED** [review]
waiting for marcin confirmation

**Comment by marcin-bazyl** [comment]
> waiting for marcin confirmation

Their fix was causing problems, so I've convinced them to revert it. It's reverted now, so this PR is good to go and can be merged.

**Review by arun3528 — APPROVED** [review]

**Pull request merged**
